### PR TITLE
Non fungible token improvements

### PR DESCRIPTION
**Pull request by YakupIpek**
This pr is at the top of #44 so it should be reviewed after #44 merged. 

This pr include workaround for State.Clear issue. 
Mint methods return token id
Token uri format improvements. 
Some small refactorings.

### CODE REVIEW

**Comment by YakupIpek** [comment]
@mrtpain @noescape00 @rowandh

**Comment by YakupIpek** [comment]
@rowandh @noescape00 @mrtpain 

This pr contains workaround for State.Clear bug and few refactorings. 

This is the NF-Token we are going to use. Others have State.Clear bug.  

I want to merge this asap if you can review and approve. Thanks

**Review by noescape00 — APPROVED** [review]
LGTM

**Comment by YakupIpek** [comment]
@mrtpain Any feedback?

File: Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs [hunk 1]
```diff
@@ -364,12 +330,12 @@ private void TransferInternal(Address to, ulong tokenId)
...
     /// <param name="from">Address from wich we want to remove the NFT.</param>
     /// <param name="tokenId">Which NFT we want to remove.</param>
-    private void RemoveNFToken(Address from, ulong tokenId)
+    private void RemoveToken(Address from, ulong tokenId)
     {
         Assert(GetIdToOwner(tokenId) == from);
         var tokenCount = GetBalance(from);
         SetBalance(from, checked(tokenCount - 1));
-        State.Clear(GetIdToOwnerKey(tokenId));
+        SetIdToOwner(tokenId, Address.Zero); //Work around for clear bug State.Clear(GetIdToOwnerKey(tokenId));
```
Comment (mrtpain) [inline Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs:338]: Nitpick - Remove this comment referencing a bug. If this is a mainnet ready contract, don't give any indication of issues or workarounds. Rather if a comment is needed just use `// clear tokenId state`

**Comment by mrtpain** [comment]
Get methods such as `BalanceOf` or `OwnerOf` probably shouldn't enforce or throw through `EnsureAddressIsNotEmpty` This can be a pain for others to integrate with these methods. 

IMO it would be better to return `0` for balanceOf and `Address.Zero` for OwnerOf rather than throw an exception. 

Idk how ETH NFTs handle this but it feels dirty to me. Up to you.

**Comment by mrtpain** [comment]
LGTM

only other comment if you wanted to match ERC721 closer would be to use `UInt256` throughout and not `ulong`

**Comment by YakupIpek** [comment]
@rowandh can you please merge it ? It is all fine. uint256 support is not belong this pr context.

**Pull request merged**